# NanoporeFiltrationCompare
Pipeline for simulation of real-time AMR detection on depth filtrated Nanopore reads, compared with results from a reference genome.
Simulation of real-time data is achieved by splitting the input Nanopore fastq file into smaller files, with sizes based on average genome coverages.
The AMR gene database used is ResFinder 4.0.
The tools used for alignment and mapping to the database is KMA.

## Requirements

- [KMA](https://bitbucket.org/genomicepidemiology/kma/src/master/) 

- Indexed ResFinder database. Created by running KMA index on [ResFinder 4.0](https://bitbucket.org/genomicepidemiology/resfinder_db/src/master/)
- phenotypes.txt file from the ResFinder database.
- Nanopore reads for each isolate.
- Paired end Illumina reads for each isolate.
- .tsv results file generated from ABRicate on the reference draft genome, for each isolate.

## Usage

To run full pipeline:
(Make sure all scripts are placed in the same directory)

```
python3 NanoporeFiltrationCompare.py [-i <Folder containing isolates>] [-o <Output folder>] [-gs <Genome size>] [-cl <Coverage list>] [-ip <Phenotypes.txt file from ResFinder database>] [-op <Output folder for plots>] [-refdb <Reference database generated by KMA index>]
```

## Input arguments
Required arguments
```
-i <Input folder. Should contain a sub-folder for each isolate. Each isolate sub-folder should contain a Nanopore fastq file, Illumina paired end fastq files (Naming have to include .R1. and .R2.), and a .tsv results file generated by running ABRicate on the reference draft genome>
-o <Path to output folder>
-gs <Genome size, used when generating smaller Nanopore fastq files. Genome size for Gram-negative bacteria is 5M>
-cl <Coverage list, used to decide the size of the smaller Nanopore fastq files, based on the required amount of reads to reach an average genome coverage with a given genome size. Values should be integers ordered from lowest to highest, eg. 1,2,3,4,5>
-ip <Path to the phenotypes.txt file from the ResFinder database. Used for genotypic to phenotypic translation>
-op <Path to output folder for plots>
-refdb <Path to reference database, generated by running KMA index on the ResFinder database>
```
Optional arguments
```
-cov <Set the minimum coverage threshold for including genes. Value between 0 and 100. Default 90.>
-id <Set the minimum identity threshold for including genes. Value between 0 and 100. Default 90.>
-md <Set the minimum depth for gene-groups to be included. Value between 0.0 and 1.0. Default 0.0>
-df <Set the minimum depth for gene-variants, based on depth relative to highest depth gene from the same gene-group. Value between 0.0 and 1.0. Default 0.0>
-sct <Skip creation of smaller Nanopore files, if they already exist. 1 to skip>
-sk <Skip KMA, if it has already been run previously. 1 to skip>
```

## Pipeline workflow

![alt text](https://github.com/CWestergaard/NanoporeFiltrationCompare/blob/master/NanoporeFiltrationCompare_Workflow.png)
